keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1004 @@
+kind: ClusterRole
+metadata:
+  name: example-com-trustmanager
+rules:
+- apiGroups:
+  - certificates.k8s.io
+  resources:
+  - signers/trustanchorsets
+  resourceNames:
+  - example.com/my-signer-name # Or example.com/* for all signers in the example.com domain.
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
+```
+
+And a component that directly reads TrustAnchorSets (typically Kubelet) would
+probably just be allowed to read all TrustAnchorSets:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: trustanchorsets-reader
+rules:
+- apiGroups:
+  - certificates.k8s.io
+  resources:
+  - trustanchorsets
+  verbs:
+  - get
+  - list
+  - watch
+```
+
+### trustAnchorSets Projected Volume Source
+
+The trustAnchorSets projected volume source is responsible for retrieving a
+designated trust anchor set from a signer and publishing it into the container
+filesystem.
+
+#### Configuration Object Definition
+
+```go
+// TrustAnchorSetsProjection is the top-level configuration object for
+// instances of the trustAnchorSets kubelet projected volume source.
+type TrustAnchorSetsProjection struct {
+  // The signer to talk to.
+  SignerName string
+
+  // Which trust anchor sets should we retrieve from the signer, and where
+  // should we write them into the volume?
+  TrustAnchorSets []TrustAnchorSetFile
+}
+
+// TrustAnchorSet is a oneof that holds different strategies for writing
+// trust anchor sets to the projected volume.
+//
+// Conceptually, it’s a oneof, but only the FlatFile option is provided
+// for now.  In the future, different formats that can hold more than one
+// named trust anchor set may be supported.  Or we might support specialty
+// formats like Java keystores.
+type TrustAnchorSetFile struct {
+  // If set, write a single trust anchor set to a single PEM file.
+  FlatFile *FlatFile
+}
+
+// FlatFile is a trust anchor set provisioning strategy that writes a single
+// named trust anchor set to a single file within the project volume.
+//
+// The file format is newline-separated PEM-wrapped DER-encoded X.509
+// certificates.
+type FlatFile struct {
+  // Which named trust anchor set should we read from the signer’s
+  // TrustAnchorSets objects?
+  Name string
+  // Where should we write the flat file in the projected volume?
+  Path string
+}
+```
+
+#### Volume Content Generation and Refresh
+
+For each trustAnchorSets volume source in the pod spec:
+
+1. Kubelet will establish a watch on the TrustAnchorSets for the named signer.
+
+2. Kubelet will wait for all relevant TrustAnchorSets objects to have the
+   "Ready" condition.
+
+3. For each FlatFile entry in the volume source:
+
+   a. Kubelet will take the union of all PEMs in `spec.sets` entries with the correct
+     `name` across all TrustAnchorSets object for the signer.
+
+   b. Kubelet will write all of the resulting certificates into a file at the designated path.
```

Comment:
Wouldn't it be simpler for the kubelet to maintain a single global cache and then pull data off of it?

Reply:
I've added a note that a single shared watch can be used.  I think it comes down to the scalability consideration --- are users likely to create a great many TrustAnchorSets?  In which case we would want to confine each Kubelet to only watch the TAS objects that are truly necessary for the workloads on the node.

I think probably users won't be creating thousands of these, though.